File: leetcode/js/count-binary-substrings.js
/**
 * @param {string} s
 * @return {number}
 */
var countBinarySubstrings = function (s) {
  let result = 0;
  let prev = 0;
  let cur = 1;

  for (let i = 1; i < s.length; i++) {
    if (s[i - 1] !== s[i]) {
      result += Math.min(prev, cur);
      prev = cur;
      cur = 1;
    } else {
      cur += 1;
    }
  }

  return result + Math.min(prev, cur);
};

File: leetcode/rust/count-binary-substrings.rs
impl Solution {
    pub fn count_binary_substrings(s: String) -> i32 {
        let mut res = 0;
        let mut prev = 0;
        let mut cur = 1;
        let chars: Vec<char> = s.chars().collect();

        for i in 1..s.len() {
            if (chars[i - 1] != chars[i]) {
                res += prev.min(cur);
                prev = cur;
                cur = 1;
            } else {
                cur += 1;
            }
        }

        res + prev.min(cur)
    }
}
